nav2_amcl/src/amcl_node.cpp

```diff
@@ -914,291 +742,262 @@ AmclNode::laserReceived(sensor_msgs::msg::LaserScan::ConstSharedPtr laser_scan)
-        odom_frame_id_);
-  } catch (tf2::TransformException e) {
-    // If we've never sent a transform, then this is normal, because the
-    // global_frame_id_ frame doesn't exist.  We only care about in-time
-    // transformation for on-the-move pose-setting, so ignoring this
-    // startup condition doesn't really cost us anything.
-    if (sent_first_transform_) {
-      RCLCPP_WARN(get_logger(), "Failed to transform initial pose in time (%s)", e.what());
-    }
-    tf2::impl::Converter<false, true>::convert(tf2::Transform::getIdentity(), tx_odom.transform);
-  }
-
-  tf2::Transform tx_odom_tf2;
-  tf2::impl::Converter<true, false>::convert(tx_odom.transform, tx_odom_tf2);
-  tf2::Transform pose_old, pose_new;
-  tf2::impl::Converter<true, false>::convert(msg.pose.pose, pose_old);
-  pose_new = pose_old * tx_odom_tf2;
-
-  // Transform into the global frame
-
-  RCLCPP_INFO(get_logger(), "Setting pose (%.6f): %.3f %.3f %.3f",
-    this->now().nanoseconds() * 1e-9,
-    pose_new.getOrigin().x(),
-    pose_new.getOrigin().y(),
-    tf2::getYaw(pose_new.getRotation()));
-  // Re-initialize the filter
-  pf_vector_t pf_init_pose_mean = pf_vector_zero();
-  pf_init_pose_mean.v[0] = pose_new.getOrigin().x();
-  pf_init_pose_mean.v[1] = pose_new.getOrigin().y();
-  pf_init_pose_mean.v[2] = tf2::getYaw(pose_new.getRotation());
-  pf_matrix_t pf_init_pose_cov = pf_matrix_zero();
-  // Copy in the covariance, converting from 6-D to 3-D
-  for (int i = 0; i < 2; i++) {
-    for (int j = 0; j < 2; j++) {
-      pf_init_pose_cov.m[i][j] = msg.pose.covariance[6 * i + j];
+      tf_broadcaster_->sendTransform(tmp_tf_stamped);
     }
   }
-
-  pf_init_pose_cov.m[2][2] = msg.pose.covariance[6 * 5 + 5];
-
-  delete initial_pose_hyp_;
-  initial_pose_hyp_ = new amcl_hyp_t();
-  initial_pose_hyp_->pf_pose_mean = pf_init_pose_mean;
-  initial_pose_hyp_->pf_pose_cov = pf_init_pose_cov;
-
-  applyInitialPose();
-}
-
-void
-AmclNode::applyInitialPose()
-{
-  std::lock_guard<std::recursive_mutex> cfl(configuration_mutex_);
-
-  // If initial_pose_hyp_ and map_ are both non-null, apply the initial
-  // pose to the particle filter state.
-
-  if (initial_pose_hyp_ != nullptr && map_ != nullptr) {
-    pf_init(pf_, initial_pose_hyp_->pf_pose_mean, initial_pose_hyp_->pf_pose_cov);
-    pf_init_ = false;
-    delete initial_pose_hyp_;
-    initial_pose_hyp_ = nullptr;
-  }
 }
 
 nav2_util::Laser *
 AmclNode::createLaserObject()
 {
-  if (map_ == NULL) {
-    RCLCPP_WARN(get_logger(), "Map is not received yet.");
-    return NULL;
-  }
+  RCLCPP_INFO(get_logger(), "createLaserObject");
+
   if (sensor_model_type_ == "beam") {
-    laser_ = new BeamModel(z_hit_, z_short_, z_max_, z_rand_, sigma_hit_, lambda_short_,
-        0.0, max_beams_, map_);
+    return new nav2_util::BeamModel(z_hit_, z_short_, z_max_, z_rand_, sigma_hit_, lambda_short_,
+             0.0, max_beams_, map_);
   } else if (sensor_model_type_ == "likelihood_field_prob") {
```

Comment:
IMHO, I think this style of returning early works better without the else. eg.

```
if (sensor_model_type == "beam)
   return x

if (sensor model type == "likelihood ..")
  return y

return z

Reply:
Yes, I agree: that's clearer. Done